[security] fixed back url verification

tessi opened the pull request:
This are the same changes @marutosi did in his PR against the `migrate/1.5_to_2.4` branch except it is posted against the `dev` branch.

Backporting to `stable` should be done once the tests ass.

see:
- http://www.redmine.org/news/90
- http://www.redmine.org/projects/redmine/wiki/Security_Advisories/13

Commit:
[security] fixed back url verification

--HG--
extra : convert_revision : svn%3Ae93f8b46-1217-0410-a6f0-8f06a7374b81/trunk%4013018

marutosi commented:
Why not for stable branch?

tessi commented:
This will probably make it into the stable branch. I'd say we "backport" (there probably not much backporting to do) this commit once the tests pass.

If it was possible, I'd put this PR to `stable` and `dev` but one can only select one target branch as far as I know :)

I've updated the PR description accordingly.

marutosi commented:
> I'd say we "backport" (there probably not much backporting to do) this commit once the tests pass.

Are you kidding?
Why do you add stable branch to .travis.yml?

> I'd put this PR to stable and dev

Are you kidding?
Do you understand what DVCS is?
Why will not you merge stable branch to dev branch?

tessi committed:
fix protocol-relative redirection test

- escape forward slash in regexp
- allow redirection to home_path '/'

tessi committed:
make url redirect tests pass

I don't know why, but the test doesn't pass when both url-tests
are in the same test-method.
When defined in separate test-methods they work fine.

tessi commented:
The tests are passing now. It turned out that the proposed fix did not escape a forward slash in the regexp and (by accident) disabled `back_url`-redirection to `'/'`.

I [proposed a diff of my changes to redmine](http://www.redmine.org/issues/16467), so we might get more eyes for review.

Furthermore, I experienced that the introduced test in [account_controller_test](https://github.com/opf/openproject/blob/7bb076fa48d840c4ba6278e344a8e434254914ac/test/functional/account_controller_test.rb#L71) fails when testing the second url of `back_urls` (weird is that the order of the urls in 'back_urls' does not matter). Splitting it into two tests magically fixed the problem, so that's what I did.

@marutosi I am sure you understand (D)VCSs far better than me - I was not kidding :) My intention was to have a single PR to handle this issue (so we have a single point for reviews and comments). But don't fear, I'm capable of bringing the changes into the `stable` branch, once review for this PR is done. Anyways, I am glad you brought this issue to my attention. Thank you.

marutosi commented:
> My intention was to have a single PR to handle this issue (so we have a single point for reviews and comments)

If you think so, why did not you send PR for stable branch?
If PR accepted in stable branch, you should merge stable branch to dev branch.

Your workflow:
- PR for dev branch
- accept
- cherry-pick revisions from dev branch to stable branch
- PR for stable branch
- accept

It is too stupid.

tessi committed:
add test for login redirecting to home_url

tessi committed:
explicitly allow home path in back_url

mfrister commented:
@marutosi Why do you think our workflow is stupid?

marutosi commented:
> Why do you think our workflow is stupid?

http://nvie.com/posts/a-successful-git-branching-model/

> Hotfix branches
> 
> They arise from the necessity to act immediately upon an undesired state of a live production  version. When a critical bug in a production version must be resolved immediately, a hotfix branch may be branched off from the corresponding tag on the master branch that marks the production version.

marutosi commented:
> Why do you think our workflow is stupid?

http://mercurial.selenic.com/wiki/MercurialDevelopmentProcess

> New features go into the main repo, whereas bugfixes are pushed into stable and pulled and merged into main. So stable is a subset of main.

linki commented on test/functional/account_controller_test.rb:
```diff
@@ -68,6 +68,21 @@ def test_login_with_logged_account
+    post :login, :username => 'jsmith', :password => 'jsmith', :back_url => '//test.foo/fake'
+    assert_redirected_to '/my/page'
+  end
+
```
the reason why the loop doesn't work is because the first #post logs you in. you might want to use the following if you want the loop back.

```
back_urls.each do |back_url|
  post :login, :username => 'jsmith', :password => 'jsmith', :back_url => back_url
  assert_redirected_to '/my/page'

  # User.current = nil has no effect
  @request.session.delete(:user_id)
end
```

linki commented on app/controllers/application_controller.rb:
```diff
@@ -427,7 +427,11 @@ def redirect_back_or_default(default)
+        # do not allow protocol relative URIs
+        if ((uri.relative? && back_url.match(%r{\A/\w})) \
+             || (uri.host == request.host) \
+             || (uri.to_s == home_path) \
```
why didn't you stick with your first implementation and used `if (uri.relative? && !back_url.match(%r{\A//}) ...`

linki committed:
using the loop requires logging the user out in-between. using the loop will make it easier to enter more examples.

linki committed:
there was a wrong parenthesis: the last match needs to be ANDed with all the prior checks. instead of changing it, refactored the code to be more clear. still allows redirects to different sub-uris.

linki commented:
fixed a small bug and refactored the code to be more clear. still allows redirects to a different sub-uri.

hschink commented on test/functional/account_controller_test.rb:
```diff
@@ -68,6 +68,25 @@ def test_login_with_logged_account
     assert_redirected_to home_url
   end
 
+  def test_login_should_not_redirect_to_another_host
```
What I don't understand is why you don't use the opportunity to make this a spec.

linki replied:
these tests were already in place and since it's a security fix we shouldn't bother with rspec conversion at this point.
